hil: Add support for tiled displays

dcz-self opened the pull request:
### Pull Request Overview

This pull request adds support for tiled displays to the screen HIL. "Tiled" as discussed in https://github.com/tock/tock/pull/3011 . The lpm013m126 is a tiled display, because the smallest unit that it can accept is a whole row of pixels. Currently, it's forced to hold a frame buffer, but after this change, it won't.

### Testing Strategy

This pull request was not tested or verified. lpm013m126 is in plans.

### TODO or Help Wanted

This pull request still needs verification and changes to the syscall driver.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

dcz-self commented on kernel/src/hil/screen.rs:
```diff
@@ -186,18 +224,41 @@ pub trait Screen {
-    /// requesting it from the screen.
-    fn get_pixel_format(&self) -> ScreenPixelFormat;
+    /// requesting it from the hardware.
+    fn get_pixel_format(&self) -> (PixelStreamFormat, Grid);
```
I'm not exactly sure where the user should get that info. Grid might change with resolution, perhaps rotation, but returning a new Grid on every call is a bit silly.

dcz-self replied:
My goal here is to push the Grid information on the user, so that they don't have an opportunity not to notice it. After all, there's no way to operate the display without knowing all of format, resolution, and grid.

The idea to put it in get_pixel_format return value came from the fact that for a screen with only one supported format, get_pixel_format *must* be called before knowing how to submit data. (So does get_resolution, but I decided by a coin flip.)

I would like to preserve that property: shove the Grid in user's face. Does this make more sense now?

The situation gets more complicated in advanced displays where the user selects the pixel format. There, the user can change the pixel format directly, wihout a call to get_pixel_format. The path to push the Grid on to the user is broken. But I think that might have a solution:

```
fn set_pixel_format(Format) -> Result<Grid, Error>
```

(I think it's enough to put it in the return value rather than a callback, the driver will probably know how to calculate the new grid.) This way, the user will see the grid before submitting data.

Do this sound like a good idea?

alexandruradovici replied:
I think the `set_pixel_format` needs to be asynchronous, as many displays will have to ask this.

dcz-self replied:
Do you mean get_pixel_format? Are there any disadvantages of making the query at initialization?

dcz-self commented on kernel/src/hil/screen.rs:
```diff
@@ -186,18 +224,41 @@ pub trait Screen {
     /// Returns the current rotation.
+    ///
+    /// The rotation is relative to a touch device, if present.
+    ///
```
I couldn't help myself mess with the rotation. Should I split it off?

alistair23 commented:
@alexandruradovici

alexandruradovici commented on kernel/src/hil/screen.rs:
```diff
@@ -116,6 +135,22 @@ impl ScreenPixelFormat {
+    /// Additional rectangles are tiled to the left and right
+    /// without overlapping, until they fill the entire display area.
+    /// Greater values are further to the right.
+    pub u16: x_offset,
+    /// Greater values are further downwards.
+    pub u16: y_offset,
```
I understand `width` and `height`, but I don't quite get the meaning of the offsets.

dcz-self replied:
I can imagine screens where the dimension is not a whole multiple of the tile size, putting the *end* at a different offset. From there, it's not a big jump to assume an odd start, perhaps due to rotation. I have no examples though.

alexandruradovici replied:
I would assume that the end would have some ignored pixels. Most probably the memory space would be there (as usually tiles are related to memory units), but the pixels are missing. I would suggest having only `width` and `height`.

dcz-self replied:
What should be done for displays rotated 180°?

phil-levis replied:
> Would there ever be a screen where the tiling didn't start at 0,0?

Yeah, I've seen this done. It happens when you want the screen to be a window into a larger framebuffer. E.g., you could do something where you load a framebuffer with a 1000x1000 image, the screen is 200x200, and you want to tile a 100x100 patch of the framebuffer 4 times into the screen.

phil-levis replied:
> What should be done for displays rotated 180°?

I think you want to operate in framebuffer space. Suppose I have a device that can detect orientation and wants to adjust the screen rotation accordingly (e.g., a modern phone). When the device rotates 180 degrees, I continue to draw up as up (the top left of the image is at 0, 0) but then I physically rotate it 180 degrees on the screen. You would also need to do geometric inversion for other inputs (e.g., a capacitive display, accelerometers), but keying this on a global rotation bit is easier than trying to dynamically switch draw calls.

phil-levis commented on kernel/src/hil/screen.rs:
```diff
@@ -212,7 +273,16 @@ pub trait Screen {
+    /// to the current pixel format and grid shape. See `get_pixel_format`.
+    /// Consecutive pixels on the screen will be set
+    /// to consecutive values from the buffer. See `PixelStreamFormat`.
+    ///
```
Generally, you should operate on a frame buffer, then apply a rotation after. Otherwise, you have to be continually transforming. I.e., always operate in the framebuffer frame of reference, not the screen frame of reference.

dcz-self replied:
Should that interpretation find itself into this change? I deliberately ignored it to limit the scope of this, and we have more semantics questions (around resolution, format, inversion) to sort out.

alexandruradovici commented:
I'll try to integrate this feedback into #3055.

dcz-self commented:
Updated: rebased, and made sure that it works (except old bugs in the lpm... driver that I'll fix in later patches).

I put this on top of the new PR: https://github.com/tock/tock/pull/3177 because the lpm... driver is broken and noncompliant to the HIL without it.

dcz-self commented:
Added userspace counterpart with an example: https://github.com/tock/libtock-rs/pull/429

alexandruradovici commented:
I suggest we fit this into #3055 which has a broader approach for displays.

dcz-self commented:
Fixing this doesn't stop #3055 from progressing. This is more or less done, so if we wait for #3055, we're just going to block further development until all of #3055 is solved.
I have improvements to the lpm013 driver already waiting, and I need tile support but don't need unification.

alexandruradovici commented:
I guess we can accept this as a temporary solution until it is merged in a similar form in #3055

bradjc commented:
Closed without prejudice. Please re-open if there are future updates.

I think changes like this need someone to reconcile a consistent view of the screen with PRs like #3811 and issue #3079.
